Jupyterlab extension for logging into a Marble node

alexandercyu opened the pull request:
Adds a button to the Jupyter Lab  toolbar that will add a cell to the top of a Jupyter notebook with prepopulated code.  The code will create and display a login interface that allows the user to choose a node to log in to and enter their login credentials.

alexandercyu committed:
Initial extension template files

alexandercyu committed:
renamed folder from jupyterlab-marble-extension to jupyterlab_marble_extension

alexandercyu committed:
Update plugin.json

Add Get Marble Session button to notebook toolbar

alexandercyu committed:
Update index.ts

- Add code cell with pre-filled code
- run the cell (currently doens't seem to be running but the return on the function runCells implies it is running)

alexandercyu committed:
Update base.css

Style input fields for node ID, username, and password

alexandercyu committed:
Update index.ts

- convert python input and password fields to jupyterlab widgets
- add submit button
- display node list as a dropdown list

alexandercyu committed:
Update index.ts

- change input fields to jupyterlab widgets
-style each input widget and align vertically
- make label for each input widget visible
- add login status widgets (Note: Not displaying properly)
- remove icon for successful login
- remove decorators for output widget capture
- remove metadata form example

alexandercyu committed:
Update index.ts

- Remove unneeded getpass import
- add text colour to input fields
- remove border (border radius not a valid style key)

alexandercyu committed:
Update index.ts

- change input field and button colours to use hex code from design
- add error message for when no choice is made for node name
- add conditions to display proper error message according to user 's dropdown choice

alexandercyu committed:
Update base.css

Remove Marble design css (moved styling to ipywidgets and Jupyter widgets)

alexandercyu committed:
Update plugin.json

Nicer indentation for json that adds button to notebook toolbar

alexandercyu committed:
Update index.ts

Remove print statements

mishaschwartz commented on .idea/workspace.xml:
don't commit this

alexandercyu replied:
".idea/" added in gitignore under Pycharm heading.

mishaschwartz commented on jupyterlab_marble_extension/.copier-answers.yml:
```diff
@@ -0,0 +1,15 @@
+# Changes here will be overwritten by Copier; NEVER EDIT MANUALLY
+_commit: v4.3.8
+_src_path: https://github.com/jupyterlab/extension-template
+author_email: test@test.com
```
add a real email here or leave it blank

alexandercyu replied:
Removed the email

mishaschwartz commented on jupyterlab_marble_extension/.github/workflows/binder-on-pr.yml:
Do we want to use binder with this? If so we should set up a github token for this action

alexandercyu replied:
I don't think binder will be needed.  It was one of the steps in the setup tutorial.

mishaschwartz commented on jupyterlab_marble_extension/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+<!-- <START NEW CHANGELOG ENTRY> -->
+
```
Add an initial entry here describing your changes

alexandercyu replied:
Added entry for not automatically installing python packages and an entry for checking for empty credentials before sending request to server.

mishaschwartz commented on jupyterlab_marble_extension/src/__tests__/jupyterlab_marble_extension.spec.ts:
```diff
@@ -0,0 +1,9 @@
+ */
+
+describe('jupyterlab-marble-extension', () => {
+  it('should be tested', () => {
```
What are some real tests that we can add here?

alexandercyu replied:
Maybe test if a cell is added to the notebook when the function is run?

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+  activate: (app: JupyterFrontEnd, notebookTracker: INotebookTracker) => {
+      console.log('JupyterLab extension jupyterlab-marble-extension is activated!');
+      const { commands } = app;
+      const command = 'jupyterlab-marble-extension:command';
```
Let's not name the command `command` since we want it to be at least a little descriptive. Pick a new name that describes the basics of what this is doing.

alexandercyu replied:
Renamed "command" to "marble_login.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+
+    // Call the command execution
+    commands.execute(command, { origin: 'init' }).catch(reason => {
+    console.error(
```
No need to call the command here. This should only be triggered by clicking the button.

alexandercyu replied:
Removed the command call.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n display(loginFailedLabelOutputWidget)'
+            )
+      }
+});
```
indent the braces properly please.

alexandercyu replied:
Indentation fixed.

mishaschwartz commented on jupyterlab_marble_extension/ui-tests/tests/jupyterlab_marble_extension.spec.ts:
```diff
@@ -0,0 +1,21 @@
+ */
+test.use({ autoGoto: false });
+
+test('should emit an activation console message', async ({ page }) => {
```
What other ui tests can we add?

alexandercyu replied:
Maybe test if a cell is added on button click?  Or test if a button has been added to the notebook toolbar?  Or test if the widgets were added?

mishaschwartz commented on node_modules/minipass-sized/node_modules/minipass/LICENSE:
node_modules should not be committed to the repo

alexandercyu replied:
"node_modules/" already in gitignore, but will look out for it in merges.

mishaschwartz commented on toolbar-button/node_modules/minipass-sized/node_modules/minipass/LICENSE:
toolbar-button code should not be committed to the repo

alexandercyu replied:
"toolbar-button" folder removed.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\ntry:' +
+            '\n from marble_client import MarbleClient' +
+            '\nexcept:' +
+            '\n !{sys.executable} -m pip install marble_client' +
```
Thinking about this again. I don't love the idea of automatically installing a package without the user explicitly agreeing. Also they might be using conda, poetry, or any other sort of package manager. Let's change this to:

```py
try:
    from marble_client import MarbleClient
except ImportError as exc:
    raise Exception("The marble_client package is required to run this cell. Please install it and run this code again.") from exc
```

alexandercyu replied:
Put in the new exception code.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+        const activeCell = notebook.activeCell;
+
+        activeCell!.model.sharedModel.setSource('import sys' +
+            '\nimport requests' +
```
`requests` may not be installed in an environment as well. Add another try/except here like below to warn the user that it is required to run this code.

alexandercyu replied:
Added a try-except with a message to the user.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\nfrom marble_client.exceptions import JupyterEnvironmentError' +
+            '\n ' +
+            '\ntry:' +
+            '\n session = MarbleClient().this_session()' +
```
instead of creating multiple MarbleClient instances throughout this code, make once instance and re-use it as needed.

alexandercyu replied:
Created a single instance -> client = MarbleClient().

Use of 'client' replaces use of 'MarbleClient()'.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n session = MarbleClient().this_session()' +
+            '\nexcept JupyterEnvironmentError:' +
+            '\n ' +
+            '\n nodeIDList = ["Node ID"]' +
```
Please use lowercase with underscores for variable names throughout. See: https://peps.python.org/pep-0008/#function-and-variable-names

alexandercyu replied:
Change function and variable names to python convention.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\nexcept JupyterEnvironmentError:' +
+            '\n ' +
+            '\n nodeIDList = ["Node ID"]' +
+            '\n selectedNode = ""' +
```
These three variables are not used

alexandercyu replied:
Removed unused variables "selected_node", "username_value", and "password_value".

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n for node in nodes:' +
+            '\n     nodeIDList.append(nodes[node].id)' +
+            '\n ' +
+            '\n nodeDropdownLabelWidget = ipywidgets.Label(value="Select the node you want to log in to:", style={"font_family":"Helvetica Neue","font_size":"16px", "text_color":"#304FFE"})' +
```
If the style value is the same for all of these three labels then store that value in a separate variable instead of copying the same dictionary 3 times.

alexandercyu replied:
Put style values in a variable.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n nodeDropdownOutput = ipywidgets.Output()' +
+            '\n ' +
+            '\n usernameLabelWidget = ipywidgets.Label(value="Enter your username:", style={"font_family":"Helvetica Neue, Helvetica","font_size":"16px", "text_color":"#304FFE"})' +
+            '\n usernameWidget = ipywidgets.Text( placeholder="", style={"description_width":"initial"}, disabled=False) ' +
```
is disabled=False necessary? Isn't that the default?

alexandercyu replied:
Removed disabled=False

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n nodeDropdownBoxWidget = ipywidgets.VBox([nodeDropdownLabelWidget, nodeDropdownWidget])' +
+            '\n nodeDropdownOutput = ipywidgets.Output()' +
+            '\n ' +
+            '\n usernameLabelWidget = ipywidgets.Label(value="Enter your username:", style={"font_family":"Helvetica Neue, Helvetica","font_size":"16px", "text_color":"#304FFE"})' +
```
should be "Enter your username or email" since emails can be used for login as well.

alexandercyu replied:
Changed label text to include "email".

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n nodeDropdownLabelWidget = ipywidgets.Label(value="Select the node you want to log in to:", style={"font_family":"Helvetica Neue","font_size":"16px", "text_color":"#304FFE"})' +
+            '\n nodeDropdownWidget = ipywidgets.Dropdown(options=nodeIDList, style={"description_width":"initial"}) ' +
+            '\n nodeDropdownBoxWidget = ipywidgets.VBox([nodeDropdownLabelWidget, nodeDropdownWidget])' +
+            '\n nodeDropdownOutput = ipywidgets.Output()' +
```
None of these output widgets are needed or used.

alexandercyu replied:
Removed output widgets and output calls in the functions.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n         if("200" in str(response)):' +
+            '\n             passwordOutput.clear_output()' +
+            '\n             with loginSuccessLabelOutputWidget:' +
+            '\n                 loginFailedLabelOutputWidget.clear_output()' +
```
Clear the outputs of **all** widgets for each case, otherwise you get multiple messages stacking up if they keep on clicking submit.

alexandercyu replied:
Output widget is cleared of outputs at each case.  Output messages no longer stack up.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n     if("selectedNode" in payload and payload["selectedNode"] != "Node ID"):' +
+            '\n         userNode = payload["selectedNode"]' +
+            '\n         url = MarbleClient()[userNode].url + "/magpie/signin" ' +
+            '\n         response = requests.post(url, headers={"Content-Type": "application/json"}, json=payload["credentials"])' +
```
This does not store the cookies in a session object that is available to the user

alexandercyu replied:
Changed to use session.post() but not sure if I'm using the correct session object.  

Using session = requests.Session() seems like it's doing a normal session not a marble client session.

Using   session = requests.Session()
            client.this_session(session)

throws a JupyterEnvironmentError

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n     userNode = "" ' +
+            '\n     response = ""' +
+            '\n     chooseAnotherNodeLabelWidget = ipywidgets.HBox([ipywidgets.Label(value="Invalid node name selected.  Please choose another node name.", style={"text_color":"#304FFE", "font_size":"16px"})]) ' +
+            '\n     loginSuccessLabelWidget = ipywidgets.HBox([ipywidgets.Label(value="Logged into " + userNode + " successfully.", style={"text_color":"green", "font_size":"16px"})]) ' +
```
userNode will never show the name of the node you're logging into on success

alexandercyu replied:
Successful login message displayed is now the one from the server response.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n         payload["credentials"]["password"] =  change["new"];' +
+            '\n ' +
+            '\n def submit(arg1):' +
+            '\n     userNode = "" ' +
```
no need to initialize variables here to the empty string

alexandercyu replied:
Removed variables that are initialized to empty string.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n         userNode = payload["selectedNode"]' +
+            '\n         url = MarbleClient()[userNode].url + "/magpie/signin" ' +
+            '\n         response = requests.post(url, headers={"Content-Type": "application/json"}, json=payload["credentials"])' +
+            '\n         if("200" in str(response)):' +
```
This is not a safe way to check for a success. The `repr` of a response object is not stable and could be anything that may or may not contain the status message. Either compare the `status_code` or `ok` attributes directly or call `raise_for_status()`

alexandercyu replied:
Changed success check to checking the response.json().

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n submitButton = ipywidgets.Button(description="Submit", disabled=False,button_style="", tooltip="Submit", icon=""' +
+            ', style={"font_family":"Helvetica Neue","font_size":"16px", "button_color":"#304FFE", "text_color":"white"} )' +
+            '\n ' +
+            '\n loginSuccessLabelOutputWidget = ipywidgets.Output()' +
```
You can just have one output widget and write the error or success message to that single output.

alexandercyu replied:
Reduced output widgets to just one and put different output messages to it.  

The success and error messages come from the response while the choose another node message is custom.

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n     with passwordOutput:' +
+            '\n         payload["credentials"]["password"] =  change["new"];' +
+            '\n ' +
+            '\n def submit(arg1):' +
```
arg1 is not used. To signal that a variable is unused in python, add an underscore before it like `_arg`

alexandercyu replied:
Added underscore to the variable name

mishaschwartz commented on jupyterlab_marble_extension/src/index.ts:
```diff
@@ -0,0 +1,161 @@
+            '\n     loginSuccessLabelWidget = ipywidgets.HBox([ipywidgets.Label(value="Logged into " + userNode + " successfully.", style={"text_color":"green", "font_size":"16px"})]) ' +
+            '\n     loginFailedLabelWidget = ipywidgets.HBox([ipywidgets.Label(value="Error Logging In", style={"text_color":"red", "font_size":"16px"})])' +
+            '\n ' +
+            '\n     if("selectedNode" in payload and payload["selectedNode"] != "Node ID"):' +
```
Why don't we also add in a check that username and password are non-empty before we submit the request so that we can tell the users that those values are required as well.

alexandercyu replied:
Added checks for username and password to check if they're non-empty.

Added error message for empty username or password.

Error messages for username, password, node name should clear correctly.

Error messages will also stack if a node name is invalid and username or password, or both, are empty.

mishaschwartz reviewed (changes requested):
All of the relevant code should be moved up one directory level: everything under the top-level `jupyterlab_marble_extensions` folder should be at the top level of the repo.

alexandercyu committed:
Deleted toolbar-button project

mishaschwartz committed:
working-example

alexandercyu committed:
Update .gitignore

Added Pycharm specific ignore

alexandercyu committed:
Update index.ts

- output widget messages no longer stack up except in the case of an invalid node name and empty user credentials
- changed request.post() to session.post()
- successful or erroneous login feedback messages is now the one from the server response
- removed variables that are initialized as empty strings
- changed success check condition to use response.json()
- removed unneeded output widgets
- added check for empty credentials

alexandercyu committed:
Update CHANGELOG.md

Add changelog items

alexandercyu committed:
Update plugin.json

Changed command name to marble_login

alexandercyu committed:
Update jupyterlab_marble_extension.spec.ts

Added test to see if a code cell is added to the top of the notebook (index zero)

alexandercyu committed:
Update index.ts

Removed command call

alexandercyu committed:
Update .copier-answers.yml

Removed email

alexandercyu committed:
Update .copier-answers.yml

Removed email

alexandercyu committed:
Deleted binder - unneeded

mishaschwartz commented:
@alexandercyu you haven't pushed any changes since my last review. If you're working on another branch you need to merge the changes here or else they won't show up on this PR.

alexandercyu committed:
Update base.css

Add blank line to satisfy prettier linting error

alexandercyu committed:
Ran prettier to fix code style issues

alexandercyu committed:
Update jupyterlab_marble_extension.spec.ts

Added tests (tests in progress)

alexandercyu committed:
Update jupyterlab_marble_extension.spec.ts

Comment out test for now (test isn't working)

alexandercyu committed:
Merge pull request #2 from DACCS-Climate/updated-example

Updated example

alexandercyu commented:
> @alexandercyu you haven't pushed any changes since my last review. If you're working on another branch you need to merge the changes here or else they won't show up on this PR.

Merged the changes from updated-example into this branch.

mishaschwartz commented on hatchling:
What is this?!

alexandercyu replied:
I don't know.  It was part of the extension template from the tutorial: https://jupyterlab.readthedocs.io/en/latest/extension/extension_tutorial.html

This line copies the template from the url: copier copy --trust https://github.com/jupyterlab/extension-template .

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\nexcept ImportError as exc:' +
+            '\n raise Exception("The marble_client package is required to run this cell. Please install it and run this code again.") from exc' +
+            '\n ' +
+            '\nfrom marble_client import MarbleClient' +
```
This is already imported above

alexandercyu replied:
Removed duplicated import.

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\n client_session = client.this_session()' +
+            '\nexcept JupyterEnvironmentError:' +
+            '\n ' +
+            '\n session = requests.Session()' +
```
This variable should be the same as in the try block

alexandercyu replied:
Changed 'client_session = client.this_session()' in the try block to 'session = client.this_session()'.

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\n password_widget =  ipywidgets.Password(style=input_field_style)' +
+            '\n password_box_widget = ipywidgets.VBox([password_label_widget, password_widget])' +
+            '\n ' +
+            '\n submit_button = ipywidgets.Button(description="Submit", button_style="", tooltip="Submit", icon=""' +
```
button_style and icon not needed

alexandercyu replied:
Removed button_style and icon from the button definition.

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\n input_field_style = {"description_width":"initial"}' +
+            '\n login_success_style = {"font_family":"Helvetica Neue","font_size":"16px", "text_color":"green"}' +
+            '\n error_style = {"font_family":"Helvetica Neue","font_size":"16px", "text_color":"red"}' +
+            '\n choose_another_node_style = {"font_family":"Helvetica Neue","font_size":"16px", "text_color":"#304FFE"}' +
```
Think about making the style into a variable to be consistent

alexandercyu replied:
Made the submit button style into a variable.

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\n ui_message_output_widget = ipywidgets.Output()' +
+            '\n ui_message_label_widget = ipywidgets.Label(value="", style=ui_label_style)' +
+            '\n ui_message_display_box_widget = ipywidgets.HBox([ui_message_label_widget])' +
+            '\n credential_error_output_widget = ipywidgets.Output()' +
```
I'd recommend just having one output and you can choose what to display in there as needed

alexandercyu replied:
Reduced to one output widget.

Removed HBox widgets for outputs.

alexandercyu committed:
Update .gitignore

Added new entries for pycharm, yarn, node_modules, and testing specific directories

mishaschwartz commented on src/index.ts:
```diff
@@ -1,110 +1,164 @@
+            '\n         response = session.post(url, headers={"Content-Type": "application/json"}, json=payload["credentials"])' +
+            '\n         response_json = response.json()' +
+            '\n ' +
+            '\n         if(response_json["code"] == 200):' +
```
https://github.com/DACCS-Climate/marble-jupyterlab-login-extension/pull/1#discussion_r2222774974

alexandercyu replied:
Now checks response.status_code for 200 response.

Also gets details of response from response.reason if there is no response.json().

alexandercyu committed:
Merge branch 'marble-extension' of https://github.com/DACCS-Climate/marble-jupyterlab-login-extension into marble-extension

alexandercyu committed:
Update index.ts

- Removed Hbox widgets for displaying output messages, just use label widgets instead.
- reduce output widgets to one
- check for server response using response.status_code instead
- if status_code is 200, then check for magpie specific json and get magpie specific response message.  Otherwise use standard response message

alexandercyu committed:
Delete hatchling

Deleted file
